src/librustc_lint/types.rs

```diff
@@ -810,51 +810,54 @@ impl LintPass for VariantSizeDifferences {
-                // sizes only make sense for non-generic types
-                let item_def_id = cx.tcx.hir.local_def_id(it.id);
-                let t = cx.tcx.type_of(item_def_id);
-                let ty = cx.tcx.erase_regions(&t);
-                let layout = cx.layout_of(ty).unwrap_or_else(|e| {
-                    bug!("failed to get layout for `{}`: {}", t, e)
-                });
-
-                if let layout::Variants::Tagged { ref variants, ref tag, .. } = layout.variants {
-                    let discr_size = tag.value.size(cx.tcx).bytes();
-
-                    debug!("enum `{}` is {} bytes large with layout:\n{:#?}",
-                      t, layout.size.bytes(), layout);
-
-                    let (largest, slargest, largest_index) = enum_definition.variants
-                        .iter()
-                        .zip(variants)
-                        .map(|(variant, variant_layout)| {
-                            // Subtract the size of the enum discriminant
-                            let bytes = variant_layout.size.bytes()
-                                .saturating_sub(discr_size);
-
-                            debug!("- variant `{}` is {} bytes large", variant.node.name, bytes);
-                            bytes
-                        })
-                        .enumerate()
-                        .fold((0, 0, 0), |(l, s, li), (idx, size)| if size > l {
-                            (size, l, idx)
-                        } else if size > s {
-                            (l, size, li)
-                        } else {
-                            (l, s, li)
-                        });
-
-                    // we only warn if the largest variant is at least thrice as large as
-                    // the second-largest.
-                    if largest > slargest * 3 && slargest > 0 {
-                        cx.span_lint(VARIANT_SIZE_DIFFERENCES,
-                                     enum_definition.variants[largest_index].span,
-                                     &format!("enum variant is more than three times larger \
-                                               ({} bytes) than the next largest",
-                                              largest));
-                    }
+        if let hir::ItemEnum(ref enum_definition, ref generics) = it.node {
+            for param in &generics.params {
+                match param.kind {
+                    hir::GenericParamKind::Lifetime { .. } => {},
+                    hir::GenericParamKind::Type { .. } => return,
+                }
+            }
+            // Sizes only make sense for non-generic types.
+            let item_def_id = cx.tcx.hir.local_def_id(it.id);
+            let t = cx.tcx.type_of(item_def_id);
+            let ty = cx.tcx.erase_regions(&t);
+            let layout = cx.layout_of(ty).unwrap_or_else(|e| {
+                bug!("failed to get layout for `{}`: {}", t, e)
```

Comment:
Or, not executing this `bug!` if the error is `Unknown`. Because if the layout depends on type parameters, it will return an `Err(Unknown(...))`. The advantage is that something like `&T` (where `T: Sized`) does *not* depend on `T`, so even more enums could be handled here.